src/ApplicationInsights.Kubernetes/PodInfoProviders/PodInfoManager.cs

```diff
@@ -0,0 +1,100 @@
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.ApplicationInsights.Kubernetes.Debugging;
+using Microsoft.ApplicationInsights.Kubernetes.Entities;
+using Microsoft.ApplicationInsights.Kubernetes.Utilities;
+
+namespace Microsoft.ApplicationInsights.Kubernetes.PodInfoProviders;
+
+internal class PodInfoManager : IPodInfoManager
+{
+    private readonly IK8sQueryClientFactory _k8SQueryClientFactory;
+    private readonly IKubeHttpClientSettingsProvider _settingsProvider;
+    private readonly IEnumerable<IPodNameProvider> _podNameProviders;
+    private readonly ApplicationInsightsKubernetesDiagnosticSource _logger = ApplicationInsightsKubernetesDiagnosticSource.Instance;
+
+    public PodInfoManager(
+        IK8sQueryClientFactory k8SQueryClientFactory,
+        IKubeHttpClientSettingsProvider settingsProvider,
+        IEnumerable<IPodNameProvider> podNameProviders)
+    {
+        _k8SQueryClientFactory = k8SQueryClientFactory ?? throw new ArgumentNullException(nameof(k8SQueryClientFactory));
+        _settingsProvider = settingsProvider ?? throw new System.ArgumentNullException(nameof(settingsProvider));
+        _podNameProviders = podNameProviders ?? throw new System.ArgumentNullException(nameof(podNameProviders));
+    }
+
+    /// <inheritdoc />
+    public async Task<K8sPod?> GetMyPodAsync(CancellationToken cancellationToken)
+    {
+        string podName = string.Empty;
+        foreach (IPodNameProvider podNameProvider in _podNameProviders)
```

Comment:
I wonder if it would be possible (and desirable) to use `Enumerable.Any()` instead of explicit loop here?

Reply:
> I wonder if it would be possible (and desirable) to use `Enumerable.Any()` instead of explicit loop here?

Yes, it is possible but I did consider it and chose not to do it. Here are my considerations:

* there is a precedence when both variables are provided
    *  I actually don't see a reason for the user to overwrite the podName by variable unless the $HOSTNAME is somehow messed up; And should that happen, user choice win.
* I guess the perf gain is very very very subtle :-)

Let me know if I miss anything.